tests for meetings/meeting and docs

chrisadubois opened the pull request.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,235 @@
+        });
+        it('calls createMeeting and returns its promise', async () => {
+          const create = spark.meetings.create('test', 'test');
+          assert.exists(create.then);
```
Nice assurance of a promise

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,235 @@
+            assert.instanceOf(meeting, Meeting, 'createMeeting should eventually resolve to a Meeting Object');
+            assert.calledOnce(spark.meetings.meetingInfo.fetchMeetingInfo);
+            assert.calledOnce(MeetingsUtil.extractDestination);
+            assert.calledOnce(MeetingsUtil.getMeetingAddedType);
```
I don't know if we need to get this granular on asserting individual steps inside of `create`. The most important thing is we assert on what the expected return value is (and that it resolves/rejects a promise).

chrisadubois replied:
I'll add 2 tests, one detailed and one contract based.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+            it('gets the reachability data instance from spark.meetings', () => {
+              const reachability = spark.meetings.getReachability();
+              assert.exists(reachability, 'reachability is defined because #setReachability has been called');
+              assert.typeOf(reachability, 'object', 'should be an object');
```
I would rather assert that specific methods exist instead of asserting it's an instanceOf or that it's an object. If we make a breaking change to the API contract, this won't help us. In this case, just asserting that there is a .getReachability() is probably enough. Then we will have separate unit tests for that component later down the line

chrisadubois replied:
yeah, i like the idea of adding assert.exists checks for the public api/event functions

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+            describe('#getMeetingByType', () => {
+              it('gets the Meeting instance from MeetingCollection using type and value', () => {
+                staticMeetings.getMeetingByType('test', 'test');
+                assert.calledOnce(MeetingCollection.getByKey);
```
Instead of asserting that in internal function was called, we should assert that `test` was returned. I would prefer to insert an actual object (multiple actually) in the `beforeEach` then assert that the correct object is returned.

chrisadubois replied:
okay, will take a look.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+              it('calls MeetingCollection to get all meetings with supplied options', () => {
+                staticMeetings.getAllMeetings({test: 'test'});
+                assert.calledOnce(MeetingCollection.getAll);
+                assert.calledWith(MeetingCollection.getAll, {test: 'test'});
```
Same as above. Insert multiple meetings in before each and assert they are all returned

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+              it('calls MeetingCollection to get all active meetings using the active option', () => {
+                staticMeetings.getActiveMeetings();
+                assert.calledOnce(MeetingCollection.getAll);
+                assert.calledWith(MeetingCollection.getAll, {active: true});
```
same as above

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+              it('calls MeetingCollection to get all scheduled meetings using the scheduled option', () => {
+                staticMeetings.getScheduledMeetings();
+                assert.calledOnce(MeetingCollection.getAll);
+                assert.calledWith(MeetingCollection.getAll, {scheduled: true});
```
same as above

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+              assert.calledOnce(onFullLocus);
+              assert.calledWith(spark.meetings.request.getActiveMeetings);
+              assert.calledWith(MeetingCollection.getByKey, 'locusUrl', 'test');
+              assert.calledWith(onFullLocus, {url: 'test'});
```
I don't think we want to assert on individual internal methods being called. We should be able to change those and as long as the end result is the same, the test should pass.

chrisadubois replied:
hmmmmm, i agree with you, but we also want to make sure when we make changes, that the function is doing everything it was initially supposed to, so if somebody changes something internal, that was initially meant to also be called (for example, unsetlocalStream AND unset localShare), we would have broken behavior.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+          done();
+        });
+      });
+      describe('#handleLocusMercury', () => {
```
looks good!

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:
```diff
@@ -0,0 +1,321 @@
+            assert.calledWith(initialSetup, 'test', {locusInfo: {initialSetup}});
+          });
+          it('should setup the meeting by difference event', async () => {
+            await spark.meetings.handleLocusEvent({locus: 'test', eventType: 'test', locusUrl: 'test'});
```
this is purely an internal handling correct? an end user wouldn't invoke that method?

chrisadubois replied:
correct

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/index.js:
```diff
@@ -0,0 +1,715 @@
+          userId: 'test',
+          resource: 'test',
+          deviceUrl: 'test',
+          locus: {url: 'test'}
```
I think it's helpful to put realistic looking data in tests. Otherwise you get a string of `test` everywhere and it's possible it's not the value you think it is but some other prop.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/index.js:
```diff
@@ -0,0 +1,715 @@
+            meeting.getMediaStreams = sinon.stub().returns(true);
+            await meeting.joinWithMedia();
+            assert.calledOnce(meeting.join);
+            assert.notCalled(meeting.getMediaStreams);
```
doesn't joinWithMedia call getMediaStreams internally? I don't know why we would assert that we don't (or why it passes)

chrisadubois replied:
likely commenting on a resolution vs rejection describe block

```
        describe('resolution', () => {
          it('should success and return a promise', async () => {
            meeting.join = sinon.stub().returns(Promise.resolve('test'));
            meeting.getMediaStreams = sinon.stub().returns(Promise.resolve(['test', 'test']));
            meeting.addMedia = sinon.stub().returns(Promise.resolve('test'));
            await meeting.joinWithMedia();
            assert.calledOnce(meeting.join);
            assert.calledOnce(meeting.getMediaStreams);
          });
        });
        describe('rejection', () => {
          it('should error out and return a promise', async () => {
            meeting.join = sinon.stub().returns(Promise.reject());
            meeting.getMediaStreams = sinon.stub().returns(true);
            await meeting.joinWithMedia();
            assert.calledOnce(meeting.join);
            assert.notCalled(meeting.getMediaStreams);
          });
        });
```

chrisadubois commented:
@andrewholsted -- yeah i definitely feel the overboardness, not arguing with you on that front.

I do feel it adds some locked-down-ness to the files, lets go over in standup tomorrow and decide as a group?

Also I added tests for MeetingCollection in a separate file that addressed your concerns for pulling values out of the Collection.

chrisadubois commented on packages/node_modules/@webex/plugin-meetings/src/meetings/collection.js:
```diff
@@ -40,6 +39,7 @@ const MeetingCollection = {
    *
    * @param {Object} options An object containg properties to filter by
    * @param {Boolean} options.active Only return active meetings
+   * @param {Boolean} options.scheduled Only return scheduled meetings
```
remove

chrisadubois commented on packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:
```diff
@@ -668,29 +696,27 @@ export default class Meeting extends StatelessSparkPlugin {
     // eslint-disable-next-line no-warning-comments
     // TODO wire into default config. There's currently an issue with the stateless plugin or how we register
-    this.mediaProperties.setMediaDirection(Object.assign(this.config.mediaSettings, options.mediaSettings));
+    this.mediaProperties.setMediaDirection(process.env.NODE_ENV === 'test' ? {} : Object.assign(this.config.mediaSettings, options.mediaSettings));
```
let's edit this with mockSpark instance so that meeting has a spark.meetings.config.mediaSettings stored due to StatelessSparkPlugin

chrisadubois commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/collection.js:
```diff
@@ -0,0 +1,69 @@
+        MeetingCollection.meetings.test = 'test';
+      });
+      it('should get the meeting', () => {
+        assert.equal(MeetingCollection.get('test'), 'test');
```
missed on the checking for size of the collection

chrisadubois commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/collection.js:
```diff
@@ -0,0 +1,69 @@
+      });
+      describe('with filters', () => {
+        beforeEach(() => {
+          MeetingCollection.meetings.test = {id: 'test', active: true};
```
change id to uuid format, don't test beyond what values are used from the meeting object

chrisadubois commented on packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/collection.js:
```diff
@@ -0,0 +1,69 @@
+    });
+    describe('#getAll', () => {
+      beforeEach(() => {
+        MeetingCollection.meetings.test = 'test';
```
using 'test' is problematic, use like id1 or id2, or value1 value2, or pmr pmr1, pmr2 etc
